### ISSUE TITLE
How to set the number of parallels when using scatter

### ISSUE THREAD

**Issue by kongxiangya** [common-workflow-language/cwltool#1827]
For example, when we run 100 samples and use scatter, we add --parallel at runtime. In reality, 100 samples will be submitted simultaneously at runtime, resulting in insufficient system memory. Is there a parameter that can set the number of allowed parallels? For example, if set to 20, only 20 samples will be run at a time. When a sample is completed, other samples will be run, and overall, the running samples will be kept to be less than or equal to 20

If this configuration is not currently available, where would you suggest adding it and how to write it appropriately

thanks

**Comment by tetron** [comment]
There isn't an option for that, although it is something that has been discussed before.

When using `cwltool --parallel` it does do basic resource management and won't run more simultaneous tasks than available RAM or cores.  So for the basic case, you might already be covered.

There's also a couple of possible versions of limiting parallelism.  You could cap the total number of simultaneous processes, or cap the fan-out of a specific scatter step.  These have different implications for scheduling and the order that things will be executed.

**Comment by kongxiangya** [comment]
> There isn't an option for that, although it is something that has been discussed before.
> 
> When using `cwltool --parallel` it does do basic resource management and won't run more simultaneous tasks than available RAM or cores. So for the basic case, you might already be covered.
> 
> There's also a couple of possible versions of limiting parallelism. You could cap the total number of simultaneous processes, or cap the fan-out of a specific scatter step. These have different implications for scheduling and the order that things will be executed.

Thank you for your answer

When using `cwltool --parallel` it does do basic resource management and won't run more simultaneous tasks than available RAM or cores.

Does this mean that the`cwltool --parallel has already considered RAM or cores when delivering tasks, and will not submit when the limit is exceeded？

You could cap the total number of simultaneous processes, or cap the fan-out of a specific scatter step.

Is this an existing configuration? Could you please provide an example to illustrate？

**Comment by mr-c** [comment]
> Does this mean that the`cwltool --parallel has already considered RAM or cores when delivering tasks, and will not submit when the limit is exceeded？

Yes, see https://github.com/common-workflow-language/cwltool/blob/16393f080b0f3cead2058be6887e3eb7b9fee849/cwltool/executors.py#L275

Note that currently, the amount of available memory and cores is only checked at the beginning. It is not updated during the workflow run.

https://github.com/common-workflow-language/cwltool/blob/16393f080b0f3cead2058be6887e3eb7b9fee849/cwltool/executors.py#L270-L271

> > You could cap the total number of simultaneous processes, or cap the fan-out of a specific scatter step.

> Is this an existing configuration? Could you please provide an example to illustrate？

No, that's not an existing configuration option. But it would be relatively straightforward to implement.

**Comment by hongduosun** [comment]
Thanks for developing cwltool! 
And do we have any plan to allow user-specified `max_cores `or `max_ram`? 
This could be straightforward to implement but is there any side effect to not doing so?
